crash when extending trait that extends jl.Enum in REPL

## Minimized code

play these statements in the repl (launched with dotr 0.25.0-RC2)

```Scala

scala> trait T extends java.lang.Enum[T]
scala> class TEnum extends T

```

## Output (click arrow to expand)
<details>

```scala
error while transforming new Enum[T]
error while transforming new Enum[T]()
error while transforming () extends Enum[T](), T {}
error while transforming class TEnum() extends Enum[T](), T {}
error while transforming () extends Object(), _root_.scala.Serializable { this: rs$line$2.type =>
  class TEnum() extends Enum[T](), T {}
}
error while transforming @scala.annotation.internal.SourceFile("rs$line$2") final module class rs$line$2$
  (
) extends Object(), _root_.scala.Serializable { this: rs$line$2.type =>
  class TEnum() extends Enum[T](), T {}
}
error while transforming package <empty> {
  final lazy module val rs$line$2: rs$line$2$ = new rs$line$2$()
  @scala.annotation.internal.SourceFile("rs$line$2") final module class 
    rs$line$2$
  () extends Object(), _root_.scala.Serializable { this: rs$line$2.type =>
    class TEnum() extends Enum[T](), T {}
  }
}
Exception in thread "main" java.lang.AssertionError: assertion failed: denotation class Enum invalid in run 2. ValidFor: Period(1..44, run = 3)
	at dotty.DottyPredef$.assertFail(DottyPredef.scala:17)
	at dotty.tools.dotc.core.Denotations$SingleDenotation.updateValidity(Denotations.scala:688)
	at dotty.tools.dotc.core.Denotations$SingleDenotation.bringForward(Denotations.scala:718)
	at dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:773)
	at dotty.tools.dotc.core.Symbols$Symbol.recomputeDenot(Symbols.scala:119)
	at dotty.tools.dotc.core.Symbols$Symbol.computeDenot(Symbols.scala:113)
	at dotty.tools.dotc.core.Symbols$Symbol.denot(Symbols.scala:106)
	at dotty.tools.dotc.core.SymDenotations$.stillValidInOwner(SymDenotations.scala:2362)
	at dotty.tools.dotc.core.SymDenotations$.stillValid(SymDenotations.scala:2358)
	at dotty.tools.dotc.core.Symbols$Symbol.isValidInCurrentRun(Symbols.scala:144)
	at dotty.tools.dotc.core.Types$NamedType.computeSymbol(Types.scala:1942)
	at dotty.tools.dotc.core.Types$NamedType.symbol(Types.scala:1937)
	at dotty.tools.dotc.core.Substituters$.subst1(Substituters.scala:27)
	at dotty.tools.dotc.core.Substituters$Subst1Map.apply(Substituters.scala:165)
	at dotty.tools.dotc.core.Types$TypeMap.mapOver$$anonfun$1(Types.scala:5003)
	at dotty.tools.dotc.core.Types$LazyRef$.apply$$anonfun$1(Types.scala:2632)
	at dotty.tools.dotc.core.Types$LazyRef.ref(Types.scala:2645)
	at dotty.tools.dotc.core.Types$TypeMap.mapOver$$anonfun$1(Types.scala:4996)
	at dotty.tools.dotc.core.Types$LazyRef$.apply$$anonfun$1(Types.scala:2632)
	at dotty.tools.dotc.core.Types$LazyRef.ref(Types.scala:2645)
	at dotty.tools.dotc.core.Types$TypeMap.mapOver$$anonfun$1(Types.scala:4996)
	at dotty.tools.dotc.core.Types$LazyRef$.apply$$anonfun$1(Types.scala:2632)
	at dotty.tools.dotc.core.Types$LazyRef.ref(Types.scala:2645)
	at dotty.tools.dotc.core.Types$TypeMap.mapOver$$anonfun$1(Types.scala:4996)
	at dotty.tools.dotc.core.Types$LazyRef$.apply$$anonfun$1(Types.scala:2632)
	at dotty.tools.dotc.core.Types$LazyRef.ref(Types.scala:2645)
	at dotty.tools.dotc.core.Types$TypeMap.mapOver$$anonfun$1(Types.scala:4996)
	at dotty.tools.dotc.core.Types$LazyRef$.apply$$anonfun$1(Types.scala:2632)
	at dotty.tools.dotc.core.Types$LazyRef.ref(Types.scala:2645)
	at dotty.tools.dotc.core.TypeComparer.secondTry$1(TypeComparer.scala:402)
	at dotty.tools.dotc.core.TypeComparer.compareNamed$1(TypeComparer.scala:274)
	at dotty.tools.dotc.core.TypeComparer.firstTry$1(TypeComparer.scala:276)
	at dotty.tools.dotc.core.TypeComparer.op$1(TypeComparer.scala:1131)
	at dotty.tools.dotc.core.TypeComparer.recur(TypeComparer.scala:1150)
	at dotty.tools.dotc.core.TypeComparer.isSubType(TypeComparer.scala:162)
	at dotty.tools.dotc.core.TypeComparer.isSubType(TypeComparer.scala:172)
	at dotty.tools.dotc.core.TypeComparer.isSubArg$1(TypeComparer.scala:1346)
	at dotty.tools.dotc.core.TypeComparer.recurArgs$1(TypeComparer.scala:1351)
	at dotty.tools.dotc.core.TypeComparer.isSubArgs(TypeComparer.scala:1354)
	at dotty.tools.dotc.core.TypeComparer.checkSubArgs$1(TypeComparer.scala:905)
	at dotty.tools.dotc.core.TypeComparer.loop$3(TypeComparer.scala:907)
	at dotty.tools.dotc.core.TypeComparer.isMatchingApply$1(TypeComparer.scala:922)
	at dotty.tools.dotc.core.TypeComparer.compareAppliedType2$1(TypeComparer.scala:1015)
	at dotty.tools.dotc.core.TypeComparer.thirdTry$1(TypeComparer.scala:498)
	at dotty.tools.dotc.core.TypeComparer.secondTry$1(TypeComparer.scala:448)
	at dotty.tools.dotc.core.TypeComparer.firstTry$1(TypeComparer.scala:345)
	at dotty.tools.dotc.core.TypeComparer.op$1(TypeComparer.scala:1131)
	at dotty.tools.dotc.core.TypeComparer.recur(TypeComparer.scala:1150)
	at dotty.tools.dotc.core.TypeComparer.isSubType(TypeComparer.scala:162)
	at dotty.tools.dotc.core.TypeComparer.tryBaseType$1(TypeComparer.scala:706)
	at dotty.tools.dotc.core.TypeComparer.compareAppliedType2$1(TypeComparer.scala:1023)
	at dotty.tools.dotc.core.TypeComparer.thirdTry$1(TypeComparer.scala:498)
	at dotty.tools.dotc.core.TypeComparer.secondTry$1(TypeComparer.scala:357)
	at dotty.tools.dotc.core.TypeComparer.firstTry$1(TypeComparer.scala:345)
	at dotty.tools.dotc.core.TypeComparer.op$1(TypeComparer.scala:1131)
	at dotty.tools.dotc.core.TypeComparer.recur(TypeComparer.scala:1150)
	at dotty.tools.dotc.core.TypeComparer.isSubType(TypeComparer.scala:162)
	at dotty.tools.dotc.core.TypeComparer.isSubType(TypeComparer.scala:172)
	at dotty.tools.dotc.core.TypeComparer.topLevelSubType(TypeComparer.scala:128)
	at dotty.tools.dotc.core.Types$Type.$less$colon$less(Types.scala:910)
	at dotty.tools.dotc.core.TypeOps$.check$1(TypeOps.scala:591)
	at dotty.tools.dotc.core.TypeOps$.checkOverlapsBounds$1(TypeOps.scala:594)
	at dotty.tools.dotc.core.TypeOps$.boundsViolations$$anonfun$2(TypeOps.scala:598)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:553)
	at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:551)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:920)
	at scala.collection.IterableOps$WithFilter.foreach(Iterable.scala:890)
	at dotty.tools.dotc.core.TypeOps$.boundsViolations(TypeOps.scala:600)
	at dotty.tools.dotc.typer.Checking$.checkBounds(Checking.scala:72)
	at dotty.tools.dotc.typer.Checking$.checkBounds(Checking.scala:83)
	at dotty.tools.dotc.transform.PostTyper$PostTyperTransformer.transform(PostTyper.scala:282)
	at dotty.tools.dotc.transform.PostTyper$PostTyperTransformer.app1$1(PostTyper.scala:262)
	at dotty.tools.dotc.transform.PostTyper$PostTyperTransformer.transform(PostTyper.scala:270)
	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform$$anonfun$2(Trees.scala:1371)
	at scala.collection.immutable.List.mapConserve(List.scala:472)
	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1371)
	at dotty.tools.dotc.transform.MacroTransform$Transformer.transform(MacroTransform.scala:58)
	at dotty.tools.dotc.transform.PostTyper$PostTyperTransformer.transform$$anonfun$7$$anonfun$1(PostTyper.scala:301)
	at dotty.tools.dotc.transform.SuperAccessors.wrapTemplate(SuperAccessors.scala:199)
	at dotty.tools.dotc.transform.PostTyper$PostTyperTransformer.transform$$anonfun$4(PostTyper.scala:301)
	at dotty.tools.dotc.transform.PostTyper$PostTyperTransformer.withNoCheckNews(PostTyper.scala:97)
	at dotty.tools.dotc.transform.PostTyper$PostTyperTransformer.transform(PostTyper.scala:302)
	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1349)
	at dotty.tools.dotc.transform.MacroTransform$Transformer.transform(MacroTransform.scala:54)
	at dotty.tools.dotc.transform.PostTyper$PostTyperTransformer.transform(PostTyper.scala:324)
	at dotty.tools.dotc.transform.MacroTransform$Transformer.transformStat$1(MacroTransform.scala:42)
	at dotty.tools.dotc.transform.MacroTransform$Transformer.transformStats$$anonfun$1(MacroTransform.scala:45)
	at dotty.tools.dotc.core.Decorators$ListDecorator$.loop$1(Decorators.scala:67)
	at dotty.tools.dotc.core.Decorators$ListDecorator$.mapconserve$extension(Decorators.scala:83)
	at dotty.tools.dotc.transform.MacroTransform$Transformer.transformStats(MacroTransform.scala:45)
	at dotty.tools.dotc.transform.MacroTransform$Transformer.transform(MacroTransform.scala:61)
	at dotty.tools.dotc.transform.PostTyper$PostTyperTransformer.transform$$anonfun$7$$anonfun$1(PostTyper.scala:301)
	at dotty.tools.dotc.transform.SuperAccessors.wrapTemplate(SuperAccessors.scala:199)
	at dotty.tools.dotc.transform.PostTyper$PostTyperTransformer.transform$$anonfun$4(PostTyper.scala:301)
	at dotty.tools.dotc.transform.PostTyper$PostTyperTransformer.withNoCheckNews(PostTyper.scala:97)
	at dotty.tools.dotc.transform.PostTyper$PostTyperTransformer.transform(PostTyper.scala:302)
	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1349)
	at dotty.tools.dotc.transform.MacroTransform$Transformer.transform(MacroTransform.scala:54)
	at dotty.tools.dotc.transform.PostTyper$PostTyperTransformer.transform(PostTyper.scala:324)
	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform$$anonfun$2(Trees.scala:1371)
	at scala.collection.immutable.List.mapConserve(List.scala:472)
	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1371)
	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transformStats(Trees.scala:1369)
	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1356)
	at dotty.tools.dotc.transform.MacroTransform$Transformer.transform(MacroTransform.scala:54)
	at dotty.tools.dotc.transform.PostTyper$PostTyperTransformer.transform(PostTyper.scala:388)
	at dotty.tools.dotc.transform.MacroTransform.run(MacroTransform.scala:21)
	at dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:295)
	at scala.collection.immutable.List.map(List.scala:246)
	at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:296)
	at dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:181)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
	at dotty.tools.dotc.Run.runPhases$5(Run.scala:191)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:199)
	at dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:64)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:206)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:148)
	at dotty.tools.repl.ReplCompiler.runCompilationUnit(ReplCompiler.scala:153)
	at dotty.tools.repl.ReplCompiler.compile(ReplCompiler.scala:163)
	at dotty.tools.repl.ReplDriver.compile(ReplDriver.scala:233)
	at dotty.tools.repl.ReplDriver.interpret(ReplDriver.scala:196)
	at dotty.tools.repl.ReplDriver.loop$1(ReplDriver.scala:129)
	at dotty.tools.repl.ReplDriver.runUntilQuit$$anonfun$1(ReplDriver.scala:132)
	at dotty.tools.repl.ReplDriver.withRedirectedOutput(ReplDriver.scala:151)
	at dotty.tools.repl.ReplDriver.runUntilQuit(ReplDriver.scala:132)
	at dotty.tools.repl.Main$.main(Main.scala:6)
	at dotty.tools.repl.Main.main(Main.scala)
```
</details>

--- a/compiler/src/dotty/tools/dotc/reporting/ErrorMessageID.scala
+++ b/compiler/src/dotty/tools/dotc/reporting/ErrorMessageID.scala
@@ -167,7 +167,8 @@ enum ErrorMessageID extends java.lang.Enum[ErrorMessageID] {
     ModifierNotAllowedForDefinitionID,
     CannotExtendJavaEnumID,
     InvalidReferenceInImplicitNotFoundAnnotationID,
-    TraitMayNotDefineNativeMethodID
+    TraitMayNotDefineNativeMethodID,
+    JavaEnumParentArgsID
 
   def errorNumber = ordinal - 2
 }

--- a/compiler/src/dotty/tools/dotc/reporting/messages.scala
+++ b/compiler/src/dotty/tools/dotc/reporting/messages.scala
@@ -1528,6 +1528,12 @@ import transform.SymUtils._
       def explain = ""
     }
 
+  class JavaEnumParentArgs(parent: Type)(using Context)
+    extends TypeMsg(JavaEnumParentArgsID) {
+      def msg = em"""not enough arguments for constructor Enum: ${hl("(name: String, ordinal: Int)")}: ${hl(parent.show)}"""
+      def explain = ""
+    }
+
   class CannotHaveSameNameAs(sym: Symbol, cls: Symbol, reason: CannotHaveSameNameAs.Reason)(using Context)
     extends SyntaxMsg(CannotHaveSameNameAsID) {
     import CannotHaveSameNameAs._

--- a/compiler/src/dotty/tools/dotc/typer/RefChecks.scala
+++ b/compiler/src/dotty/tools/dotc/typer/RefChecks.scala
@@ -95,7 +95,7 @@ object RefChecks {
    *  and required classes. Also check that only `enum` constructs extend
    *  `java.lang.Enum`.
    */
-  private def checkParents(cls: Symbol)(using Context): Unit = cls.info match {
+  private def checkParents(cls: Symbol, parentTrees: List[Tree])(using Context): Unit = cls.info match {
     case cinfo: ClassInfo =>
       def checkSelfConforms(other: ClassSymbol, category: String, relation: String) = {
         val otherSelf = other.declaredSelfTypeAsSeenFrom(cls.thisType)
@@ -109,12 +109,25 @@ object RefChecks {
       for (reqd <- cinfo.cls.givenSelfType.classSymbols)
         checkSelfConforms(reqd, "missing requirement", "required")
 
+      def illegalEnumFlags = !cls.isOneOf(Enum | Trait)
+      def isJavaEnum = parents.exists(_.classSymbol == defn.JavaEnumClass)
+
       // Prevent wrong `extends` of java.lang.Enum
-      if !migrateTo3 &&
-         !cls.isOneOf(Enum | Trait) &&
-         parents.exists(_.classSymbol == defn.JavaEnumClass)
-      then
+      if !migrateTo3 && illegalEnumFlags && isJavaEnum then
         report.error(CannotExtendJavaEnum(cls), cls.sourcePos)
+      else if illegalEnumFlags && isJavaEnum then
+        val javaEnumCtor = defn.JavaEnumClass.primaryConstructor
+        parentTrees.exists(parent =>
+          parent.tpe.typeSymbol == defn.JavaEnumClass
+          && (
+            parent match
+              case tpd.Apply(tpd.TypeApply(fn, _), _) if fn.tpe.termSymbol eq javaEnumCtor =>
+                // here we are simulating the error for missing arguments to a constructor.
+                report.error(JavaEnumParentArgs(parent.tpe), cls.sourcePos)
+                true
+              case _ =>
+                false
+          ))
 
     case _ =>
   }
@@ -1089,7 +1102,7 @@ class RefChecks extends MiniPhase { thisPhase =>
   override def transformTemplate(tree: Template)(using Context): Tree = try {
     val cls = ctx.owner.asClass
     checkOverloadedRestrictions(cls)
-    checkParents(cls)
+    checkParents(cls, tree.parents)
     if (cls.is(Trait)) tree.parents.foreach(checkParentPrefix(cls, _))
     checkCompanionNameClashes(cls)
     checkAllOverrides(cls)

--- a/compiler/src/dotty/tools/dotc/typer/RefChecks.scala
+++ b/compiler/src/dotty/tools/dotc/typer/RefChecks.scala
@@ -109,25 +109,26 @@ object RefChecks {
       for (reqd <- cinfo.cls.givenSelfType.classSymbols)
         checkSelfConforms(reqd, "missing requirement", "required")
 
-      def illegalEnumFlags = !cls.isOneOf(Enum | Trait)
-      def isJavaEnum = parents.exists(_.classSymbol == defn.JavaEnumClass)
+      def isClassExtendingJavaEnum =
+        !cls.isOneOf(Enum | Trait) && parents.exists(_.classSymbol == defn.JavaEnumClass)
 
       // Prevent wrong `extends` of java.lang.Enum
-      if !migrateTo3 && illegalEnumFlags && isJavaEnum then
-        report.error(CannotExtendJavaEnum(cls), cls.sourcePos)
-      else if illegalEnumFlags && isJavaEnum then
-        val javaEnumCtor = defn.JavaEnumClass.primaryConstructor
-        parentTrees.exists(parent =>
-          parent.tpe.typeSymbol == defn.JavaEnumClass
-          && (
-            parent match
-              case tpd.Apply(tpd.TypeApply(fn, _), _) if fn.tpe.termSymbol eq javaEnumCtor =>
-                // here we are simulating the error for missing arguments to a constructor.
-                report.error(JavaEnumParentArgs(parent.tpe), cls.sourcePos)
-                true
-              case _ =>
-                false
-          ))
+      if isClassExtendingJavaEnum then
+        if !migrateTo3 then // always error, only traits or enum-syntax is possible under scala 3.x
+          report.error(CannotExtendJavaEnum(cls), cls.sourcePos)
+        else
+          // conditionally error, we allow classes to extend java.lang.Enum in scala 2 migration mode,
+          // however the no-arg constructor is forbidden, we must look at the parent trees to see
+          // which overload is called.
+          val javaEnumCtor = defn.JavaEnumClass.primaryConstructor
+          parentTrees.exists {
+            case parent @ tpd.Apply(tpd.TypeApply(fn, _), _) if fn.tpe.termSymbol eq javaEnumCtor =>
+              // here we are simulating the error for missing arguments to a constructor.
+              report.error(JavaEnumParentArgs(parent.tpe), cls.sourcePos)
+              true
+            case _ =>
+              false
+          }
 
     case _ =>
   }

--- a/tests/neg/extend-java-enum-migration.check
+++ b/tests/neg/extend-java-enum-migration.check
@@ -10,3 +10,7 @@
 14 |class Sub extends T // error
    |      ^
    |      not enough arguments for constructor Enum: (name: String, ordinal: Int): Enum[T]
+-- [E160] Type Error: tests/neg/extend-java-enum-migration.scala:17:10 -------------------------------------------------
+17 |val foo = new java.lang.Enum[Color] {} // error
+   |          ^
+   |          not enough arguments for constructor Enum: (name: String, ordinal: Int): Enum[Color]

--- a/tests/neg/extend-java-enum-migration.check
+++ b/tests/neg/extend-java-enum-migration.check
@@ -1,0 +1,12 @@
+-- [E160] Type Error: tests/neg/extend-java-enum-migration.scala:9:12 --------------------------------------------------
+9 |final class C extends jl.Enum[C] // error
+  |            ^
+  |            not enough arguments for constructor Enum: (name: String, ordinal: Int): Enum[C]
+-- [E160] Type Error: tests/neg/extend-java-enum-migration.scala:11:7 --------------------------------------------------
+11 |object O extends jl.Enum[O.type] // error
+   |       ^
+   |       not enough arguments for constructor Enum: (name: String, ordinal: Int): Enum[O.type]
+-- [E160] Type Error: tests/neg/extend-java-enum-migration.scala:14:6 --------------------------------------------------
+14 |class Sub extends T // error
+   |      ^
+   |      not enough arguments for constructor Enum: (name: String, ordinal: Int): Enum[T]

--- a/tests/neg/extend-java-enum-migration.scala
+++ b/tests/neg/extend-java-enum-migration.scala
@@ -1,0 +1,14 @@
+import java.{lang => jl}
+
+import language.`3.0-migration`
+
+// This file is different from `tests/neg/extend-java-enum.scala` as we
+// are testing that it is illegal to *not* pass arguments to jl.Enum
+// in 3.0-migration
+
+final class C extends jl.Enum[C] // error
+
+object O extends jl.Enum[O.type] // error
+
+trait T extends jl.Enum[T] // ok
+class Sub extends T // error

--- a/tests/neg/extend-java-enum-migration.scala
+++ b/tests/neg/extend-java-enum-migration.scala
@@ -12,3 +12,6 @@ object O extends jl.Enum[O.type] // error
 
 trait T extends jl.Enum[T] // ok
 class Sub extends T // error
+
+abstract class Color(name: String, ordinal: Int) extends java.lang.Enum[Color](name, ordinal) // ok
+val foo = new java.lang.Enum[Color] {} // error

